docs/bsp.md

```diff
@@ -697,6 +716,106 @@ trait ResourcesItem {
 }
 ```
 
+### Task Notifications
+
+The server may inform the client on the state of running tasks, such as compilation or tests.
+When beginning a task, the server may send `build/taskStart`, intermediate updates may be sent in
+`build/taskProgress`.
+
+If a `build/taskStart` notification has been sent, the server must send `build/taskFinish`
+on completion of the same task.
+
+Conversely, a `build/taskFinish` notification may be sent even if `build/taskStart` was not sent.
+
+`build/taskStart`, `build/taskProgress` and `build/taskFinished` notifications for the same task must use the same `taskId`.
+
+Tasks that are spawned by another task should reference the originating task's `taskId` in their own `taskId`'s
+`parent` field. Tasks spawned directly by a request should reference the request's `originId` parent.
+
+#### Task Started
+
+Notification:
+
+* method: `build/taskStart`
+* params: `TaskStartParams` defined as follows:
+
+```scala
+trait TaskStartParams {
+    /** Unique id of the task with reference to parent task id */
+    def taskId: TaskId
+
+    /** Timestamp of the event in milliseconds. */
+    def eventTime: Long
+
+    /** Message describing the task. */
+    def message: String
+
+    /** Optional metadata about the task.
+      * For compilation tasks, this should be a CompileTask object.
+      * For test tasks, this should be a TestTask object. */
+    def data: Option[Json]
+}
+```
+
+#### Task Progress
+
+After a `taskStart` and before `taskFinished` for a `taskId`, the server may send any number of progress notifications.
+
+* method: `build/taskProgress`
+* params: `TaskProgressParams` defined as follows:
+
+```scala
+trait TaskProgressParams {
+    /** Unique id of the task with reference to parent task id */
+    def taskId: TaskId
+
+    /** Timestamp of the event in milliseconds. */
+    def eventTime: Long
+
+    /** Message describing the task. */
+    def message: String
+
+    /** If known, total amount of work units in this task. */
+    def total: Option[Long]
+
+    /** If known, completed amount of work units in this task. */
+    progress: Option[Long]
+
+    /** Name of a work unit. For example, "files" or "tests". May be empty. */
+    unit: String
+
+    /** Optional metadata about the task progress. */
+    data: Option[Json]
```

Comment:
Instead of `data: Option[Json]`, how about adding optional fields for test and compile reports?

```scala
def compileReport: Option[CompileReport]
def testReport: Option[TestReport]
```

That would simplify the serialization/deserialization.